### PR TITLE
Add user agent to storage requests for telemetry fixes #680

### LINKED ISSUE #680: add unique user agent to all storage requests to be able to get logs server side
to monitor and get insights about customer trends in using storage related operations (listening on blobs/ queues - blob outputs) and basically all storage requests that we control, we need to send a unique user agent header. This way we can collect metrics and logs from Storage server side.

### PR DESCRIPTION
**Pull request by shobak101**
checked with Storage client team and there is no way to set this on the whole client, so added the call for list blobs and get queue messages

### CODE REVIEW

**Comment by fabiocav** [comment]
@brettsam I'll take a look at this, but could you help and provide some feedback?

File: src/Microsoft.Azure.WebJobs.Storage/OperationContextHelper.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            context.SendingRequest += (sender, e) =>
+            {
+                e.Request.UserAgent += userAgent;
+            };
+
+            return context;
+        }
+
+        /// <summary>Returns a default operation context.</summary>
+        public static OperationContext GetDefaultOperationContext()
```
Comment (fabiocav) [inline src/Microsoft.Azure.WebJobs.Storage/OperationContextHelper.cs]: Doesn't look like this method is used anywhere. Is it needed?

**Comment by fabiocav** [comment]
Added a minor comment about an unused method, but looks good.

Have you considered using the `GlobalSendingRequest` event? If we want to have that header sent with every request made to storage, would that be a better option (and also, potentially prevent issues down the road where we may forget to call the helper to add the event handler)?

File: src/Microsoft.Azure.WebJobs.Storage/OperationContextHelper.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.WindowsAzure.Storage;
+
+#if PUBLICSTORAGE
+namespace Microsoft.Azure.WebJobs.Storage
+#else
+namespace Microsoft.Azure.WebJobs.Host.Storage
+#endif
```
Comment (davidebbo) [inline src/Microsoft.Azure.WebJobs.Storage/OperationContextHelper.cs]: Why the conditional namespace?

File: src/Microsoft.Azure.WebJobs.Storage/OperationContextHelper.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+{
+    internal class OperationContextHelper
+    {
+        /// <summary>Adds handler to manipulate User Agent for Requests</summary>
+        public static OperationContext AddWebJobsUserAgentToOperationContext(OperationContext context)
+        {
+            var userAgent = "-AzureWebJobs";
```
Comment (davidebbo) [inline src/Microsoft.Azure.WebJobs.Storage/OperationContextHelper.cs]: I'd just inline this where it's used.

File: src/Microsoft.Azure.WebJobs.Storage/Blob/StorageBlobClient.cs [hunk 4]
```diff
@@ -58,6 +58,9 @@ public Task<ServiceProperties> GetServicePropertiesAsync(CancellationToken cance
             BlobListingDetails blobListingDetails, int? maxResults, BlobContinuationToken currentToken,
             BlobRequestOptions options, OperationContext operationContext, CancellationToken cancellationToken)
         {
+            // Add WebJobs to user agent to make it easier to collect usage information from Storage telemetry
+            operationContext = OperationContextHelper.AddWebJobsUserAgentToOperationContext(operationContext);
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Storage/Blob/StorageBlobClient.cs]: >operationContext = OperationContextHelper.AddWebJobsUserAgentToOperationContext(operationContext); [](start = 12, length = 98)

Could you make this an extension method so the call was something like 'operationContext.AddWebJobsUserAgent();'?

**Comment by brettsam** [comment]
I was just about to ask whether we'd need to do this in Extensions as well (Timer uses storage, for example). If GlobalSendingRequest would handle that for every storage operation coming from the function app, that would be nice.

---
In reply to: [235451342](https://github.com/Azure/azure-webjobs-sdk/pull/777#issuecomment-235451342) [](ancestors = 235451342)

**Comment by shobak101** [comment]
@brettsam I just tested Fabio's suggestion and it works, all calls to storage have this -AzureWebJobs added to the useragent, will send a new PR now with the latest changes

**Comment by fabiocav** [comment]
Looks good. A minor thing that might be helpful is to add a comment to the `Initialize` method body to explain that it is a noop and exists to expose a convenient way to trigger the static initialization. That might give others some context and help avoid confusion (or the temptation to remove it :) ) 
Aside from that, good to :shipit: !

**Comment by shobak101** [comment]
added the comment and pushed to master

**Pull request closed without merging**